[No QA] issue:10289 Speed Up Tests Run Time

mzohaibqc opened the pull request:
Added github matrix for jest unit tests
Added action to run lint on changed files

<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

@rushatgabhane 
@puneetlath 

### Details
<!-- Explanation of the change or anything fishy that is going on -->
I added github matrix to distribute tests on multiple github runner instances to reduce time.

### Fixed Issues
<!---
Please replace GH_LINK with the link to the GitHub issue this Pull Request is fixing.
Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the issue; otherwise, the linking will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/10289

Do NOT only link the issue number like this: $ #<number-of-the-issue>
--->
$ https://github.com/Expensify/App/issues/10289

### Tests
<!---
Add a numbered list of manual tests you performed that validates your changes work on all platforms, and that there are no regressions present.
Add any additional test steps if test steps are unique to a particular platform.
Manual test steps should be written so that your reviewer can repeat and verify one or more expected outcomes in the development environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

- [ ] Verify that no errors appear in the JS console

### PR Review Checklist
<!--
This is a checklist for PR authors & reviewers. Please make sure to complete all tasks and check them off once you do, or else Expensify has the right not to merge your PR!
-->
#### PR Author Checklist
- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] iOS / native
    - [x] Android / native
    - [x] iOS / Safari
    - [x] Android / Chrome
    - [x] MacOS / Chrome
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by tagging the marketing team on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.
</details>

### QA Steps
<!---
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

- [ ] Verify that no errors appear in the JS console

### Screenshots
<!-- Add screenshots for all platforms tested. Pull requests won't be merged unless the screenshots show the app was tested on all platforms.-->

#### Web
<!-- Insert screenshots of your changes on the web platform-->

#### Mobile Web
<!-- Insert screenshots of your changes on the web platform (from a mobile browser)-->

#### Desktop
<!-- Insert screenshots of your changes on the desktop platform-->

#### iOS
<!-- Insert screenshots of your changes on the iOS platform-->

#### Android
<!-- Insert screenshots of your changes on the Android platform-->

mzohaibqc committed:
issue: 10289 Added github matrix for jest unit tests
Added action to run lint on changed files

mzohaibqc commented:
I have read the CLA Document and I hereby sign the CLA

mzohaibqc commented:
@puneetlath @rushatgabhane Kindly approve workflow so all steps can run and be verified. This is a work in progress.

rushatgabhane reviewed (changes requested):
@mzohaibqc I have some questions, and I'm curious where is the config for **caching**?

mzohaibqc commented on .github/workflows/lint.yml:
```diff
@@ -12,7 +12,16 @@ jobs:
       - uses: Expensify/App/.github/actions/composite/setupNode@main
-
-      - run: npm run lint
+      - uses: tj-actions/eslint-changed-files@v13.3
```
I found that lint step is taking 2 minutes to complete and we should also consider reducing this time so I changed this to run lint on only changed files. My intention is good here to reduce the time but if you have any concerns, I can revert this.

mzohaibqc replied:
That's great idea. I will do it separately.

mzohaibqc commented on .github/workflows/preDeploy.yml:
```diff
@@ -9,7 +9,36 @@ jobs:
+
+        - name: Jest Unit Tests
+          if: ${{ matrix.chunk < 3 }}
+          run: npx jest --listTests --json | jq -cM '[_nwise(length / 3 | ceil)]' | jq '.[${{ matrix.chunk }}] | .[] | @text' | xargs yarn test
```
Just bear with me. It's not final code. Caching is not implemented yet.

mzohaibqc commented on .github/workflows/preDeploy.yml:
```diff
@@ -9,7 +9,36 @@ jobs:
+        # If automatic signing is enabled, iOS builds will fail, so ensure we always have the proper profile specified
+        - name: Check Provisioning Style
+          run: |
+            if grep -q 'PROVISIONING_PROFILE_SPECIFIER = chat_expensify_appstore' ios/NewExpensify.xcodeproj/project.pbxproj; then
```
these steps were part of test sub action
`
uses: Expensify/App/.github/workflows/test.yml@main
`
and there is limitation that we can't use matrix in nested/reusable actions which I needed so I have to move that code here in main file.

mzohaibqc committed:
Removed linting changes
Fixed chunk number comparison

mzohaibqc commented:
@puneetlath Kindly approve this worflow to run.

mzohaibqc committed:
Added build matrix for test job to reduce time

mzohaibqc committed:
Merge branch 'Expensify:main' into issues/10289

mzohaibqc committed:
Fixed action yml lint/syntax issues

mzohaibqc commented:
> @mzohaibqc you can verify the workflow locally too.
> 
> 1. npm run gh-actions-build
> 2. npm run gh-actions-validate

@rushatgabhane Thanks man :)

I fixed all issues and updated PR. But everytime I need someone's approval to test what are imrovements over the previous setup. I've added all the steps, added cache for jest and added github matrix to reduce the time. I noticed that `loginTest.js` is taking the most time to run so I executed that into separate runner and splitted remaining tests into 2 runner jobs.

There was an if condition which I commented out for now which I need you or someone else to discuss and understand the logic behind that so then add that with right changes

```
if: ${{ github.actor != 'OSBotify' || github.event_name == 'workflow_call' }}
```
I think ` github.event_name == 'workflow_call'` will be removed since it not `workflow_call` anymore but need your guidance to correctly implement this condition.

mzohaibqc commented:
@rushatgabhane @puneetlath Kindly review the PR.

mzohaibqc committed:
Fixed concurrency count
removed yarn and used npm test to execute tests

mzohaibqc commented on .github/workflows/preDeploy.yml:
```diff
@@ -9,7 +9,36 @@ jobs:
     uses: Expensify/App/.github/workflows/lint.yml@main
 
```
@rushatgabhane I have verified all scripts by running these commands. Kindly review this PR. I have done my changes. 

@puneetlath Kindly review this PR

mzohaibqc commented:
@puneetlath @rushatgabhane Still waiting for review.

mzohaibqc commented on .github/workflows/preDeploy.yml:
```diff
@@ -9,7 +9,46 @@ jobs:
+      # If automatic signing is enabled, iOS builds will fail, so ensure we always have the proper profile specified
+      - name: Check Provisioning Style
+        run: |
+          if grep -q 'PROVISIONING_PROFILE_SPECIFIER = chat_expensify_appstore' ios/NewExpensify.xcodeproj/project.pbxproj; then
```
@rushatgabhane these checks were part of reusable workflow for unit tests. I simply copied that here.  don't know what does this mean.

moved this 

https://github.com/Expensify/App/blob/main/.github/workflows/test.yml#L11

to here

mzohaibqc commented on package.json:
```diff
@@ -215,12 +215,13 @@
     "trailingComma": "all"
   }
-}
+}
```
sure, will add

mzohaibqc commented:
> @mzohaibqc bump ^ [#10617 (comment)](https://github.com/Expensify/App/pull/10617#discussion_r963100322)

Added replies

rushatgabhane reviewed (commented):
@mzohaibqc why aren't we making all the changes in [test.yml](https://github.com/Expensify/App/blob/main/.github/workflows/test.yml)?


Please make all the changes in `test.yml` for cleaner code.

mzohaibqc commented on .github/workflows/preDeploy.yml:
```diff
@@ -9,7 +9,46 @@ jobs:
+        if: ${{ fromJSON(matrix.chunk) < 2 }}
+        run: npx jest --listTests --json --testPathIgnorePatterns=loginTest.js | jq -cM '[_nwise(length / 2 | ceil)]' | jq '.[${{ matrix.chunk }}] | .[] | @text' | xargs npm test
+      
+      - name: Login Unit Tests
```
@rushatgabhane  Sure, I will add a comment in code

loginTest takes the most time, almost same time that is required for all other tests. It blocks other tests execution causing a lot of time. However if we run login test in separate runner and all other tests in another runner, both will run in parallel and will complete almost togather or whichever will take more time. This will reduce test time to almost half.

mzohaibqc commented:
> @mzohaibqc why aren't we making all the changes in [test.yml](https://github.com/Expensify/App/blob/main/.github/workflows/test.yml)?
> 
> Please make all the changes in `test.yml` for cleaner code.

@rushatgabhane  I am using Github Build Matrix which is not supported for reusable workflows e.g. test.yml


> @mzohaibqc why aren't we making all the changes in [test.yml](https://github.com/Expensify/App/blob/main/.github/workflows/test.yml)?
> 
> Please make all the changes in `test.yml` for cleaner code.
@rushatgabhane 

Nested/resuable workflow does not support matrix which I am using to run tests in parallel. If this is deal-breaker then I can add changes in `test.yml` file. We might need to duplicate some code for setting up Node and for `Check Provisioning Style` step (I don't know what it does and why it's in test.yml).

rushatgabhane commented:
@parasharrajat here's @mzohaibqc's answer
https://github.com/Expensify/App/pull/10617#issuecomment-1246016106

mzohaibqc commented:
> Why are we making changes to `PreDeploy.yml` workflow? How is this related to PR checks?

@parasharrajat can we have a short meeting to discuss and resolve this. Let me know your availability. We can have a zoom call, a very short one (5 - 10 min).

mzohaibqc commented:
> Why are we making changes to `PreDeploy.yml` workflow? How is this related to PR checks?

@parasharrajat Thanks for raising this question. Now I noticed that test.yml is invoked as reusable workflow from preDeploy as well as it runs on pull_request event.

I will fix this and will remove `preDeploy.yml` changes.

mzohaibqc committed:
Removed changes from preDeploy.yml file and added in test.yml

mzohaibqc committed:
Added missing newline in package.json

mzohaibqc commented:
@parasharrajat Kindly review now. I removed all changes in `preDeploy.yml file` and added build matrix in `test.yml` to reduce tests execution time.

parasharrajat reviewed (changes requested):
I understood the concept of matrix and the way you have applied it here. But I am concerned about the approach. 
1. It is not scalable. If I change the file name it will break. 
2. High maintenance, we will have to change the workflow to accommodate new changes or when new tests are added. 
3. You haven't explained in the PR description, why we need to build a matrix. Also, why 4 parallel instances?

How can we solve 1 & 2?

In addition,
1. Jest cache should be configured for all workflows which are running tests.

mzohaibqc commented on .github/workflows/test.yml:
```diff
@@ -10,23 +10,46 @@ jobs:
             echo "Error: Automatic provisioning style is not allowed!"
             exit 1
           fi
-
-      - name: Jest Unit Tests
-        run: npm run test
-        env:
-          CI: true
-
-      - name: Pull Request Tests
-        run: tests/unit/getPullRequestsMergedBetweenTest.sh
```
I moved it to a job which is taking minimum time and if it will break, it will be shown as failed step but I agree with your comment that it will not take any time so I can move it to top.

mzohaibqc commented on .github/workflows/test.yml:
```diff
@@ -10,23 +10,46 @@ jobs:
+      
       # If automatic signing is enabled, iOS builds will fail, so ensure we always have the proper profile specified
       - name: Check Provisioning Style
+        if: ${{ fromJSON(matrix.chunk) == 3 }}
```
I agree, will revert this change

mzohaibqc commented on .github/workflows/test.yml:
```diff
@@ -10,23 +10,46 @@ jobs:
+    strategy:
+      fail-fast: false
+      matrix:
+        chunk: [0, 1, 2, 3]
```
I am using 2 chunks  (0 and 1) for splitting tests and being used as index in `jq -cM '[_nwise(length / 2 | ceil)]'`
```
npx jest --listTests --json --testPathIgnorePatterns=loginTest.js | jq -cM '[_nwise(length / 2 | ceil)]' | jq '.[${{ matrix.chunk }}] | .[] | @text' | xargs npm test
```

mzohaibqc commented on .github/workflows/test.yml:
```diff
@@ -10,23 +10,46 @@ jobs:
     runs-on: ubuntu-latest
+    name: test (job ${{ matrix.chunk }})
+    env:
+      CI: true
```
We need this for tests as it's being used earlier but I can move it back to command level instead for whole job.

mzohaibqc commented:
> I understood the concept of matrix and the way you have applied it here. But I am concerned about the approach.
> 
> 1. It is not scalable. If I change the file name it will break.
> 2. High maintenance, we will have to change the workflow to accommodate new changes or when new tests are added.
> 3. You haven't explained in the PR description, why we need to build a matrix. Also, why 4 parallel instances?
> 
> How can we solve 1 & 2?
> 
> In addition,
> 
> 1. Jest cache should be configured for all workflows which are running tests.

1. If that's the case and it's highly possible then we can run all tests including loginTest.js or any name in future in parallel jobs without specifying a separate job for 1 file. We can increase the number of jobs to reduce time.
2. Splitting all tests into `n` parallel jobs/runners will be a one time change and when more tests files will be added, files will be dvided within all jobs and no code change will be required.

About Jest cache

Any workflow that is running tests, is using `test.yml` as reusable workflow e.g. `preDeploy.yml` will invoke `test.yml` to run jest tests so we don't need to add cache in parent workflows.

mzohaibqc committed:
Added code to make this scalable

mzohaibqc commented:
@parasharrajat Kindly review the changes. I added some chnages to make it more scalable and easy to maintain. If we need more or less number of parallel job, we just need to update environment variable `CHUNKS` and everything else will work.

mzohaibqc commented on .github/workflows/test.yml:
```diff
@@ -23,10 +47,19 @@ jobs:
+
+      - name: All Unit Tests
+        if: ${{ matrix.chunk < env.CHUNKS }}
+        run: npx jest --listTests --json | jq -cM '[_nwise(length / ${{ env.CHUNKS }} | ceil)]' | jq '.[${{ matrix.chunk }}] | .[] | @text' | xargs npm test
```
tried locally and it happens if `matrix.chunk` is greater than last index i.e. 2 in this case (0, 1, 2)
But this line should not execute for matrix.chunk=3 since it's not less than total CHUNKS i.e. 3

xargs npm test will result into following command for chunk 0

```
jest "/mnt/c/Users/Zohaib/Projects/App/tests/unit/loginTest.js" "/mnt/c/Users/Zohaib/Projects/App/tests/unit/ReportUtilsTest.js" "/mnt/c/Users/Zohaib/Projects/App/tests/unit/fetchFreePlanVerifiedBankAccountTest.js" "/mnt/c/Users/Zohaib/Projects/App/tests/unit/MigrationTest.js" "/mnt/c/Users/Zohaib/Projects/App/tests/unit/awaitStagingDeploysTest.js" "/mnt/c/Users/Zohaib/Projects/App/tests/unit/createOrUpdateStagingDeployTest.js" "/mnt/c/Users/Zohaib/Projects/App/tests/unit/checkDeployBlockersTest.js" "/mnt/c/Users/Zohaib/Projects/App/tests/unit/NetworkTest.js" "/mnt/c/Users/Zohaib/Projects/App/tests/unit/RequestTest.js" "/mnt/c/Users/Zohaib/Projects/App/tests/unit/WalletTest.js"
```

mzohaibqc replied:
@parasharrajat  Are you running locally since this workflow never ran on github action since it needs approval

![image](https://user-images.githubusercontent.com/13637803/191788028-20fda1ae-3d77-4f83-856b-557e762a2fbd.png)

mzohaibqc replied:
@parasharrajat It's because you are creating 1 chunk ([[all files paths]]) and then accessing index 1 which is not there.

Based on code, `matrix.chunk` can never be equal to `env.CHUNKS` for this command since it's protected by `if` condition
```
- name: All Unit Tests
        if: ${{ matrix.chunk < env.CHUNKS }}
```

mzohaibqc replied:
And running pull request tests for last iteration
```
- name: Pull Request Tests
        # Pull request related tests will be run in separate runner in parallel.
        if: ${{ matrix.chunk == env.CHUNKS }}
        run: tests/unit/getPullRequestsMergedBetweenTest.sh
```

parasharrajat replied:
Oh, I see. I will recheck

mzohaibqc committed:
Added comments

mzohaibqc commented on .github/workflows/test.yml:
```diff
@@ -6,10 +6,34 @@ on:
+        uses: actions/github-script@v6
+        with:
+          script: |
+            core.setOutput('MATRIX', Array.from({ length: process.env.CHUNKS + 1 }, (v, i) => i));
```
Added comment

parasharrajat commented:
@puneetlath Could you please approve the workflows/checks on this PR/

parasharrajat commented:
Thanks for doing that. @mzohaibqc Why there are 30 jobs running here? I thought it will be 3.

parasharrajat commented on .github/workflows/test.yml:
```diff
@@ -6,10 +6,36 @@ on:
 jobs:
+  config:
+    runs-on: ubuntu-latest
+    name: Define matrix parameters
+    outputs:
+      MATRIX: ${{ steps.set-matrix.outputs.MATRIX }}
+    steps:
+      - name: Set Matrix
+        id: set-matrix
+        uses: actions/github-script@v6
+        with:
+          # Generate matrix array i.e. [0, 1, 2, ...., CHUNKS] for test job
+          script: |
+            core.setOutput('MATRIX', Array.from({ length: process.env.CHUNKS + 1 }, (v, i) => i));
```
I see that we are using this run to define a number of chunks which seems like completely unnecessary work in GitHub runner. 

Is there a way to get the number of matrices from the configuration in jest or somewhere in test files without this step?

mzohaibqc replied:
We are doing this to reduce time in CI so it make sense to add here in respective action. Also, there is no such configuration in jest to run test in chunks and define some parameter.

parasharrajat replied:
> We are doing this to reduce time in CI so it make sense to add here in respective action

I agree but the problem with this is that it will be fragile. As soon as someone modifies a test file and a time-consuming test case to it, it will start consuming more time. 


Now, let's say that the user wants to optimize the execution time of tests. what are his options? 
1. Modify the test files order so that time-consuming tests are grouped in separate chunck. 
2. Break tests in subsets. 
4. Modify the action file to increase the chunks. 

I don't think either of these is very ideal. It is overhead. 

Instead, if there is a way to configure some kind of groups and these run parallel in its matrix then the user will just change the configuration and it will all be handled automatically. 

One way of achieving this could be the use of groups. https://www.npmjs.com/package/jest-runner-groups. 

Assign each test to a group. While running the tests via CI get the number of groups and create a matrix of length equal to # groups. 

What do you thing/

parasharrajat replied:
But first, let's get rid of the `Set Matrix` step and use a static matrix. I don't see any advantage of this step. 

The way user is setting `chunk: n` env, they can update the matrix : [0, 2, 3, ..n].

parasharrajat replied:
What are your thoughts?

mzohaibqc replied:
> But first, let's get rid of the `Set Matrix` step and use a static matrix. I don't see any advantage of this step.
> 
> The way user is setting `chunk: n` env, they can update the matrix : [0, 2, 3, ..n].

@parasharrajat  Only advantage to use this steps and common `env.CHUNKS` is that we don't have magic numbers all over and f someone forgets to update those numbers in commands and just updates the matrix, it will break/fail.

mzohaibqc replied:
> > We are doing this to reduce time in CI so it make sense to add here in respective action
> 
> I agree but the problem with this is that it will be fragile. As soon as someone modifies a test file and a time-consuming test case to it, it will start consuming more time.
> 
> Now, let's say that the user wants to optimize the execution time of tests. what are his options?
> 
> 1. Modify the test files order so that time-consuming tests are grouped in separate chunk.
> 2. Break tests in subsets.
> 3. Modify the action file to increase the chunks.
> 
> I don't think either of these is very ideal. It is overhead.
> 
> Instead, if there is a way to configure some kind of groups and these run parallel in its matrix then the user will just change the configuration and it will all be handled automatically.
> 
> One way of achieving this could be the use of groups. https://www.npmjs.com/package/jest-runner-groups.
> 
> Assign each test to a group. While running the tests via CI get the number of groups and create a matrix of length equal to # groups.
> 
> What do you thing?

We can do this but it has more maintainance cost as developer need to put new test files in specific group or new groups. They have to first check which group is taking most time and then adjust it by changing group name.

Also, using this package, one can assign more than one group (intentionally or unintentionally) to single test file, causing same file being run/executes by multiple runners.

But if you strongly feel it's better than what we have, I can do that.

parasharrajat replied:
Ok. Thanks. I think we can drop this for now.

mzohaibqc replied:
@parasharrajat  If everything is good, can you kindly add your review checklist so all steps go green.

parasharrajat commented on .github/workflows/test.yml:
```diff
@@ -6,10 +6,36 @@ on:
 jobs:
+  config:
+    runs-on: ubuntu-latest
+    name: Define matrix parameters
+    outputs:
+      MATRIX: ${{ steps.set-matrix.outputs.MATRIX }}
+    steps:
+      - name: Set Matrix
+        id: set-matrix
+        uses: actions/github-script@v6
+        with:
+          # Generate matrix array i.e. [0, 1, 2, ...., CHUNKS] for test job
+          script: |
+            core.setOutput('MATRIX', Array.from({ length: process.env.CHUNKS + 1 }, (v, i) => i));
```
`process.env.CHUNKS` is the string to due to concatenation it will give 31 chunks answer to my previous question.

mzohaibqc replied:
I will fix this

mzohaibqc replied:
@parasharrajat Fixed and pushed code. Kindly have a look

parasharrajat replied:
Thanks. let's discuss the other thread.

mzohaibqc committed:
Fixed chunking, adding code to parse chunk numbers

parasharrajat commented:
@puneetlath Could you please trigger checks?

puneetlath commented:
Triggered.

parasharrajat commented on .github/workflows/test.yml:
```diff
@@ -6,10 +6,36 @@ on:
+        with:
+          # Generate matrix array i.e. [0, 1, 2, ...., CHUNKS] for test job
+          script: |
+            core.setOutput('MATRIX', Array.from({ length: Number(process.env.CHUNKS) + 1 }, (v, i) => i));
```
```suggestion
            core.setOutput('MATRIX', Array.from({ length: Number(process.env.CHUNKS) }, (v, i) => i));
```

So based on the latest test runs, four jobs are running. 
![image](https://user-images.githubusercontent.com/24370807/194394461-f0ef7933-3769-4a27-bf38-62c05efcad10.png)

So it would be good if only the number of jobs equal to the chunk value is triggered. I think you just need to remove `+ 1`.

mzohaibqc replied:
So if you want to create `n` jobs if `CHUNKS` value is `n` then jest based test will run in `n - 1` jobs since we need one job for  `Pull Request Tests`. Then that command will change from 
```
- name: Pull Request Tests
        # Pull request related tests will be run in separate runner in parallel.
        if: ${{ fromJSON(matrix.chunk) == fromJSON(env.CHUNKS) }}
        run: tests/unit/getPullRequestsMergedBetweenTest.sh
```

to 

```
- name: Pull Request Tests
        # Pull request related tests will be run in separate runner in parallel.
        if: ${{ fromJSON(matrix.chunk) == fromJSON(env.CHUNKS) - 1 }}
        run: tests/unit/getPullRequestsMergedBetweenTest.sh
```

And for jest based test, command will change from 
```
- name: All Unit Tests
        if: ${{ fromJSON(matrix.chunk) < fromJSON(env.CHUNKS) }}
         run: npx jest --listTests --json | jq -cM '[_nwise(length / ${{ fromJSON(env.CHUNKS) }} | ceil)]' | jq '.[${{ fromJSON(matrix.chunk) }}] | .[] | @text' | xargs npm test
```

to 

```
- name: All Unit Tests
        if: ${{ fromJSON(matrix.chunk) < fromJSON(env.CHUNKS) - 1 }}
        # Split the jest based test files in multiple chunks/groups and then execute them in parallel in different jobs/runners.
        run: npx jest --listTests --json | jq -cM '[_nwise(length / ${{ fromJSON(env.CHUNKS) }} | ceil)]' | jq '.[${{ fromJSON(matrix.chunk) }}] | .[] | @text' | xargs npm test
```


so for `n` equals `4`

matrix will be `[0, 1, 2, 3]` and jest based tests will run for `0`, `1`, `2` and for index `3`, pull request test will be run.

Let me know if that sound good then I can add this change.

parasharrajat replied:
So I don't get this bug change. I was thinking if we just change the matrix creation logic, it will be fine. 

So 
```
chunks: 3
Current matrix logic : [0 , 1, 2, 3]
My suggestion: change it to  [0, 1, 2] (n-1) last element

then change the name to start from 1 instead of 0
```

parasharrajat replied:
Never mind, It seems you are correct. I didn't notice that you said `so for n equals 4`. 

Please make the necessary changes.

parasharrajat replied:
Is this change commited? I don't see this. Jobs should be equal to # of chunks. For our case, 3 jobs for 3 chunks.

mzohaibqc replied:
> Never mind, It seems you are correct. I didn't notice that you said `so for n equals 4`.
> 
> Please make the necessary changes.

I though you agreed with me. 

We in any case, we need to handle `n - 1` senario since we are using index for arrays of chunks. 

If you want 3 jobs for `CHUNKS: 3` then I've to add changes mentioned above like
```
jq -cM '[_nwise(length / ${{ fromJSON(env.CHUNKS) - 1 }}

if: ${{ fromJSON(matrix.chunk) == fromJSON(env.CHUNKS) - 1 }}

 if: ${{ fromJSON(matrix.chunk) < fromJSON(env.CHUNKS) - 1 }}
````

parasharrajat replied:
yeah sure.

mzohaibqc replied:
Pushed changes

now CHUNKS: 3
Jest based tests will split into Job 1 and Job 2
Job 3 will execute PR tests

parasharrajat reviewed (commented):
Waiting for the change. I will approve it as soon as done. Thanks

parasharrajat commented on .github/workflows/test.yml:
```diff
@@ -6,10 +6,36 @@ on:
+    needs: config
     if: ${{ github.actor != 'OSBotify' || github.event_name == 'workflow_call' }}
     runs-on: ubuntu-latest
+    name: test (job ${{ matrix.chunk }})
```
```suggestion
    name: test (job ${{ matrix.chunk + 1 }})
```
Make sure to convert it to number if `matrix.chunk` is string here.

mzohaibqc replied:
@parasharrajat I pushed this change
```
name: test (job ${{ fromJSON(matrix.chunk) + 1 }})
```

Do you want ant other change? If I missed any requested changes, do let me know, I will do that

mzohaibqc committed:
Added code to start job name from job 1 instead of job 0

mzohaibqc committed:
Addressed feedback, chnaged number of jobs from CHUNKS + 1 to CHUNKS

parasharrajat commented:
Thanks. Please do this
1. Prefix title with `[No QA]`

parasharrajat reviewed (approved):
LGTM. 

cc: @puneetlath Please trigger the checks before appoving. I want to see the result. 

#### PR Reviewer Checklist
- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] iOS / native
    - [x] Android / native
    - [x] iOS / Safari
    - [x] Android / Chrome
    - [x] MacOS / Chrome
    - [x] MacOS / Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by tagging the marketing team on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

:ribbon: :eyes: :ribbon:  C+ reviewed

parasharrajat reviewed (changes requested):
Thanks. Please do this

- Prefix title with `[No QA]`

mzohaibqc commented:
> 1. [No QA]

@parasharrajat  I did not get that, which title I need to modify?

mzohaibqc commented:
> > 1. [No QA]
> 
> @parasharrajat I did not get that, which title I need to modify?

I updated PR title

parasharrajat commented:
Thanks

mzohaibqc commented:
@parasharrajat Validate Github action is causing this error/earning although it's a valid syntax.

![image](https://user-images.githubusercontent.com/13637803/194615742-98061f94-9312-4bff-a532-1470f4bce25f.png)

I was using this logic as a js script because added linting mechanism for Github actions does not accepts this. 
![image](https://user-images.githubusercontent.com/13637803/194616047-dbca5c7e-0a74-4d93-8c59-555cb39da404.png)

i.e.
I was using 
```
core.setOutput('MATRIX', Array.from({ length: Number(process.env.CHUNKS) + 1 }, (v, i) => i));
```
instead of this
```
core.setOutput('MATRIX', Array.from({ length: Number(process.env.CHUNKS) }, (v, i) => i));
```

parasharrajat commented on .github/workflows/test.yml:
```diff
@@ -6,10 +6,36 @@ on:
+    needs: config
     if: ${{ github.actor != 'OSBotify' || github.event_name == 'workflow_call' }}
     runs-on: ubuntu-latest
+    name: test (job ${{ fromJSON(matrix.chunk) + 1 }})
```
Error is being thrown here. I think it should be
```suggestion
    name: test (job ${{ Number(matrix.chunk) + 1 }})
```

mzohaibqc replied:
@parasharrajat  I tried `Number()` locally and ran github action validate command and it is failing again. Linting has some parsing issue. Also, I don't think `Number` will work in github action expression. It's not metioned in available functions for expressions. 

I tried this in my own test repository and it works. IMHO, it's issue with Linting tool.

mzohaibqc replied:
Linting has issues with `+` sign

workflows/test.yml:31:46: got unexpected character '+' while lexing expression, expecting 'a'..'z', 'A'..'Z', '_', '0'..'9', ''', '}', '(', ')', '[', ']', '.', '!', '<', '>', '=', '&', '|', '*', ',', ' ' [expression]

mzohaibqc replied:
Other option is to create another variable like `JEST_CHUNKS` or something that you suggest and we can calculate and set in github js script where we are setting `MATRIX` environment variable for matrix config.
```
script: |
            core.setOutput('MATRIX', Array.from({ length: Number(process.env.CHUNKS) }, (v, i) => i));
            core.setOutput('JEST_CHUNKS', Number(process.env.CHUNKS) - 1);



- name: All Unit Tests
        if: ${{ fromJSON(matrix.chunk) < fromJSON(env.JEST_CHUNKS) }}
        run: npx jest --listTests --json | jq -cM '[_nwise(length / ${{ fromJSON(env.JEST_CHUNKS) }} | ceil)]' | jq '.[${{ fromJSON(matrix.chunk) }}] | .[] | @text' | xargs npm test

      - name: Pull Request Tests
        if: ${{ fromJSON(matrix.chunk) == fromJSON(env.JEST_CHUNKS) }}
        run: tests/unit/getPullRequestsMergedBetweenTest.sh
```

mzohaibqc replied:
and even if we do the above, we can't start job name from 1 i.e. `job 1` because we again have to add 1 to make it right and using `+` is not allowed in linting rule. Tried different methods locally but did not work.

parasharrajat replied:
Let me check this

parasharrajat replied:
Ok, `+` is not a valid operator in Github actions. 

May be we can start with this Matrix and then job name will be correct and adjust the other operations to meet this. 
```yaml
core.setOutput('MATRIX', Array.from({ length: Number(process.env.CHUNKS) }, (v, i) => i  +1));
```

If this is not working, then `JEST_CHUNKS` approach is fine too.

mzohaibqc replied:
Here is output for above command

```
[
  [],
  [  
    "/mnt/c/Users/BIG BANG/OneDrive/Desktop/Upwork/App/tests/unit/loginTest.js",
    "/mnt/c/Users/BIG BANG/OneDrive/Desktop/Upwork/App/tests/unit/ReportUtilsTest.js",
    "/mnt/c/Users/BIG BANG/OneDrive/Desktop/Upwork/App/tests/unit/fetchFreePlanVerifiedBankAccountTest.js",
    "/mnt/c/Users/BIG BANG/OneDrive/Desktop/Upwork/App/tests/unit/createOrUpdateStagingDeployTest.js",
    "/mnt/c/Users/BIG BANG/OneDrive/Desktop/Upwork/App/tests/unit/checkDeployBlockersTest.js",
    "/mnt/c/Users/BIG BANG/OneDrive/Desktop/Upwork/App/tests/unit/awaitStagingDeploysTest.js",
    "/mnt/c/Users/BIG BANG/OneDrive/Desktop/Upwork/App/tests/unit/MigrationTest.js",
    "/mnt/c/Users/BIG BANG/OneDrive/Desktop/Upwork/App/tests/unit/NetworkTest.js",
    "/mnt/c/Users/BIG BANG/OneDrive/Desktop/Upwork/App/tests/unit/RequestTest.js",
    "/mnt/c/Users/BIG BANG/OneDrive/Desktop/Upwork/App/tests/unit/WalletTest.js",
    "/mnt/c/Users/BIG BANG/OneDrive/Desktop/Upwork/App/tests/unit/OptionsListUtilsTest.js",
    "/mnt/c/Users/BIG BANG/OneDrive/Desktop/Upwork/App/tests/actions/SessionTest.js",
    "/mnt/c/Users/BIG BANG/OneDrive/Desktop/Upwork/App/tests/actions/ReportTest.js",
    "/mnt/c/Users/BIG BANG/OneDrive/Desktop/Upwork/App/tests/unit/EmojiRegexTest.js",
    "/mnt/c/Users/BIG BANG/OneDrive/Desktop/Upwork/App/tests/unit/isStagingDeployLockedTest.js"
  ],
  [
    "/mnt/c/Users/BIG BANG/OneDrive/Desktop/Upwork/App/tests/unit/isPullRequestMergeableTest.js",
    "/mnt/c/Users/BIG BANG/OneDrive/Desktop/Upwork/App/tests/unit/TranslateTest.js",
    "/mnt/c/Users/BIG BANG/OneDrive/Desktop/Upwork/App/tests/unit/GithubUtilsTest.js",
    "/mnt/c/Users/BIG BANG/OneDrive/Desktop/Upwork/App/tests/actions/ReimbursementAccountTest.js",
    "/mnt/c/Users/BIG BANG/OneDrive/Desktop/Upwork/App/tests/unit/PaymentUtilsTest.js",
    "/mnt/c/Users/BIG BANG/OneDrive/Desktop/Upwork/App/tests/unit/nativeVersionUpdaterTest.js",
    "/mnt/c/Users/BIG BANG/OneDrive/Desktop/Upwork/App/tests/unit/enhanceParametersTest.js",
    "/mnt/c/Users/BIG BANG/OneDrive/Desktop/Upwork/App/tests/unit/versionUpdaterTest.js",
    "/mnt/c/Users/BIG BANG/OneDrive/Desktop/Upwork/App/tests/unit/GooglePlacesUtilsTest.js",
    "/mnt/c/Users/BIG BANG/OneDrive/Desktop/Upwork/App/tests/unit/CurrencySymbolUtilsTest.js",
    "/mnt/c/Users/BIG BANG/OneDrive/Desktop/Upwork/App/tests/unit/NumberUtilsTest.js",
    "/mnt/c/Users/BIG BANG/OneDrive/Desktop/Upwork/App/tests/unit/LocalePhoneNumberTest.js",
    "/mnt/c/Users/BIG BANG/OneDrive/Desktop/Upwork/App/tests/unit/GitUtilsTest.js",
    "/mnt/c/Users/BIG BANG/OneDrive/Desktop/Upwork/App/tests/unit/CardUtilsTest.js"
  ]
]
```

mzohaibqc replied:
IMO, we should increase CHUNKS to 4 at-least or 5 otherwise there will be negligible gain in execution time because with CHUNKS: 3, jest test will be divided into two jobs.

parasharrajat replied:
So there will be a total of 3 jobs. One is for Pull request tests. If this is taking less than 5 minutes in total then I guess two is fine too. let's see when checks are triggered.

parasharrajat commented on .github/workflows/test.yml:
```diff
@@ -23,10 +49,20 @@ jobs:
+          key: ${{ runner.os }}-jest
+
+      - name: All Unit Tests
+        if: ${{ fromJSON(matrix.chunk) < (fromJSON(env.CHUNKS) - 1) }}
```
```suggestion
        if: ${{ Number(matrix.chunk) < (Number(env.CHUNKS) - 1) }}
```

parasharrajat commented on .github/workflows/test.yml:
```diff
@@ -23,10 +49,20 @@ jobs:
+      - name: All Unit Tests
+        if: ${{ fromJSON(matrix.chunk) < (fromJSON(env.CHUNKS) - 1) }}
+        # Split the jest based test files in multiple chunks/groups and then execute them in parallel in different jobs/runners.
+        run: npx jest --listTests --json | jq -cM '[_nwise(length / ${{ fromJSON(env.CHUNKS) - 1 }} | ceil)]' | jq '.[${{ fromJSON(matrix.chunk) }}] | .[] | @text' | xargs npm test
```
```suggestion
        run: npx jest --listTests --json | jq -cM '[_nwise(length / ${{ Number(env.CHUNKS) - 1 }} | ceil)]' | jq '.[${{ Number(matrix.chunk) }}] | .[] | @text' | xargs npm test
```

parasharrajat commented on .github/workflows/test.yml:
```diff
@@ -23,10 +49,20 @@ jobs:
 
       - name: Pull Request Tests
+        # Pull request related tests will be run in separate runner in parallel.
+        if: ${{ fromJSON(matrix.chunk) == (fromJSON(env.CHUNKS) - 1) }}
```
```suggestion
        if: ${{ Number(matrix.chunk) == (Number(env.CHUNKS) - 1) }}
```

parasharrajat commented:
Please validate the actions in local before pushing the changes.

mzohaibqc committed:
Fixed github action linting issues

parasharrajat commented:
Thanks, Looking good to me. @puneetlath Please trigger the checks to confirm.

parasharrajat commented:
cc: @puneetlath This is ready for final review.

mzohaibqc commented:
@parasharrajat Can you kindly help me why PR Author Checklist action is failing and can you kindly review PR description section (markdown) and see if I am missing something

mzohaibqc commented:
@puneetlath  I copied updated PR Author checklist from here
https://raw.githubusercontent.com/Expensify/App/main/.github/PULL_REQUEST_TEMPLATE.md

and updated my PR

Kindly trigger the failed action.

parasharrajat commented:
You removed this 
![image](https://user-images.githubusercontent.com/24370807/194920895-63cc28ad-5169-4ed7-a2cc-cb6fc31303bb.png)
and this
![image](https://user-images.githubusercontent.com/24370807/194920936-a0f50a97-4b9a-4121-af82-edf996c00ae4.png)

I think reverting those things will do the trick

parasharrajat commented:
May be a recheck will fix it.

mzohaibqc commented:
> May be a recheck will fix it.

thanks, it worked.

mzohaibqc commented:
@puneetlath @parasharrajat Can you kindly approve this.

AndrewGable commented:
No need to tag us @mzohaibqc, we will review the PR in a timely manner 😃

puneetlath reviewed (approved):
Seems to make sense to me. Tapping in @AndrewGable for a second review since he's more familiar with these tests.

puneetlath commented:
@parasharrajat C+ reviewer checklist test is failing.

parasharrajat commented:
All good now.

parasharrajat commented:
In short, this PR
1. Converts the single test workflow to 4 parallel jobs.

AndrewGable reviewed (approved):
Nice one!

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/3230780415) to staging by @AndrewGable in version: 1.2.13-0 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/3239913870) to production by @yuwenmemon in version: 1.2.13-5 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅

roryabraham reviewed (commented):
Also, I wonder if this would be simpler?

```ym;    
jobs:
  test:
    if: ${{ github.actor != 'OSBotify' || github.event_name == 'workflow_call' }}
    runs-on: ubuntu-latest
    strategy:
      matrix:
        fail-fast: false
        chunk: [1, 2, 3]
    name: test (job ${{ matrix.chunk }})
    env:
      CI: true
    steps:
      ...
      ...
```

i.e: There's no need for the `config` job?

roryabraham commented:
What verifications did we do to see if this actually speeds anything up? Asking because it adds a fair amount of complexity to the test workflow and if it doesn't come with really meaningful improvements in turnaround time then I wonder if it's really worth it?

mzohaibqc commented:
> Also, I wonder if this would be simpler?
> 
> ```ym;    
> jobs:
>   test:
>     if: ${{ github.actor != 'OSBotify' || github.event_name == 'workflow_call' }}
>     runs-on: ubuntu-latest
>     strategy:
>       matrix:
>         fail-fast: false
>         chunk: [1, 2, 3]
>     name: test (job ${{ matrix.chunk }})
>     env:
>       CI: true
>     steps:
>       ...
>       ...
> ```
> 
> i.e: There's no need for the `config` job?

@roryabraham this cobfig job is to reduce changes required if we plan to increase test execution parallelism in future then it will be just one line change. Otherwise one has to change matrix and all the logic for splitting tests into right number of jobs.

mzohaibqc commented:
> What verifications did we do to see if this actually speeds anything up? Asking because it adds a fair amount of complexity to the test workflow and if it doesn't come with really meaningful improvements in turnaround time then I wonder if it's really worth it?
> 

@roryabraham i suggested to run login file test into a separate job since that's bottleneck to reduce execution time like all other tests take same or less time what login test file takes but this idea was rejected.

roryabraham commented:
> Otherwise one has to change matrix and all the logic for splitting tests into right number of jobs

I don't see how this is true 😕

AndrewGable commented:
The tests seem faster to me, but I didn't personally time them. Open to @roryabraham's suggestions for fixes and I can implement them.

roryabraham commented:
I'm all for the caching, less keen on the matrix approach. It results in 3 separate npm installs in the 3 separate runners, which is a bit of a bottleneck. I've seen the tests be complete in one of the matrix runner while a slower npm install is happening in another

parasharrajat commented:
I didn't notice that it is running `npm install` in all parallel jobs. That is kind of bad.

roryabraham commented:
The more we live with this, the less I like the matrix approach for splitting up tests. It makes it harder to see where errors are coming from, since `xargs` masks the exit code of other commands:

<img width="451" alt="image" src="https://user-images.githubusercontent.com/47436092/208170634-64f734a7-c660-4ec8-b384-838ac32907d0.png">

AndrewGable commented:
Let's just remove the matrix if it's causing more problems then value, I am not even sure if this PR accomplished the original goal of the issue in the end.

roryabraham commented:
I was running into [some difficulties](https://expensify.slack.com/archives/C01GTK53T8Q/p1671217362571039) after upgrading Jest, and had to refactor the `test.yml` workflow to get things passing. You can see that PR [here](https://github.com/Expensify/App/pull/12168). It's simpler than what we have here, but overall I still wouldn't be opposed to taking it one step further and removing the matrix if it doesn't really have a big impact.

> this cobfig job is to reduce changes required if we plan to increase test execution parallelism in future then it will be just one line change. Otherwise one has to change matrix and all the logic for splitting tests into right number of jobs

With the refactor in https://github.com/Expensify/App/pull/12168, you can still adjust the number of chunks with a simple 1-line change

puneetlath commented:
Created an issue for removing it here: https://github.com/Expensify/App/issues/14088
